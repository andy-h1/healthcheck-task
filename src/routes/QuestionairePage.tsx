import JSONData from "../data/questionaire.json";
import useQuestionNavigation from "../hooks/useQuestionNavigation";
import useCalculateOutcome from "../hooks/useCalculateOutcome";
import { QuestionaireType } from "../types/questionaire";
import ProgressBar from "../components/ProgressBar";
import Summary from "../components/Summary";
import Questionaire from "../components/Questionaire";
import Header from "../components/Header";
import { useContext } from "react";
import { QuestionaireContext } from "../context/QuestionaireContext";

const QuestionairePage = (): React.ReactNode => {
  const data = JSONData as QuestionaireType;
  const { questions, outcomes } = data;
  const initialQuestionId = questions?.[0]?.id;
  const context = useContext(QuestionaireContext);

  console.log({ context });

  const {
    currentQuestionId,
    currentScore,
    selectedAnswerId,
    isAnswerSelected,
    currentQuestion,
    handleBackClick,
    handleNextClick,
    handleSelectAnswerClick,
    handleRestart
  } = useQuestionNavigation(questions, initialQuestionId);

  const outcome = useCalculateOutcome(questions, currentScore, outcomes);

  const findCurrentQuestionIndex = questions.findIndex(
    (q) => q.id === currentQuestionId
  );

  const progress = Math.round(
    (findCurrentQuestionIndex / (questions.length - 1)) * 100
  );

  const allQuestionsAnswered =
    findCurrentQuestionIndex === questions.length - 1;

  return (
    <div className="flex h-screen w-screen items-center justify-center md:bg-teal-300">
      <div className="flex h-dvh w-full max-w-screen-sm flex-col items-center justify-between bg-white px-4 py-4 md:h-4/5 md:rounded-md md:border md:border-gray-300 ">
        <div className="w-full">
          <Header
            findCurrentQuestionIndex={findCurrentQuestionIndex}
            allQuestionsAnswered={allQuestionsAnswered}
            onBack={handleBackClick}
          />
          <ProgressBar progress={progress} />
        </div>

        {!allQuestionsAnswered && currentQuestion && (
          <Questionaire
            question={currentQuestion}
            selectedAnswerId={selectedAnswerId}
            isAnswerSelected={isAnswerSelected}
            onAnswerSelect={handleSelectAnswerClick}
            onNext={handleNextClick}
          />
        )}

        {allQuestionsAnswered && outcome && (
          <Summary outcome={outcome} onRestart={handleRestart} />
        )}
      </div>
    </div>
  );
};

export default QuestionairePage;
