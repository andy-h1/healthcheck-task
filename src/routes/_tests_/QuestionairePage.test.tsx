import { render, screen, waitFor } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { describe, expect, test, vi } from "vitest";
import QuestionairePage from "../QuestionairePage";
import {
  mockQuestions,
  mockInitialQuestionId
} from "../../test/utils/mockData";

let mockState = {
  currentQuestion: mockQuestions.find((q) => q.id === mockInitialQuestionId),
  selectedAnswerId: "",
  isAnswerSelected: false
};

const handleSelectAnswerClick = vi.fn((answerId) => {
  mockState = {
    ...mockState,
    selectedAnswerId: answerId,
    isAnswerSelected: true
  };
});

const handleNextClick = vi.fn();

vi.mock("../../hooks/useQuestionNavigation", () => ({
  default: vi.fn(() => ({
    ...mockState,
    handleSelectAnswerClick,
    handleNextClick
  }))
}));

describe("QuestionairePage", () => {
  test("it should render", () => {
    const { container } = render(<QuestionairePage />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test("user can select an answer which enables them to click the next button", async () => {
    const { rerender } = render(<QuestionairePage />);

    expect(mockState.isAnswerSelected).toBe(false);

    const yesButton = screen.getByRole("button", { name: /yes/i });
    userEvent.click(yesButton);

    await waitFor(() => {
      expect(handleSelectAnswerClick).toHaveBeenCalled();
      expect(mockState.isAnswerSelected).toBe(true);
    });

    // Have to rerender the component to update the state of the next button
    rerender(<QuestionairePage />);

    const nextButton = screen.getByRole("button", { name: /next/i });

    expect(nextButton).toBeEnabled();

    userEvent.click(nextButton);

    await waitFor(() => {
      expect(handleNextClick).toHaveBeenCalled();
    });
  });
});
