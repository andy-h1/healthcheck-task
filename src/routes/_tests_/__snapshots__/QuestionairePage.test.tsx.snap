// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QuestionairePage > it should render 1`] = `
<div
  class="flex h-screen w-screen items-center justify-center md:bg-teal-300"
>
  <div
    class="flex h-dvh w-full max-w-screen-sm flex-col items-center justify-between bg-white px-4 py-4 md:h-4/5 md:rounded-md md:border md:border-gray-300 "
  >
    <div
      class="w-full"
    >
      <span
        class="mb-4 flex items-center"
      >
        <h1
          class="grow text-center"
        >
          Heartburn Checker
        </h1>
      </span>
      <div
        class="overflow-hidden rounded-xl bg-gray-200"
      >
        <div
          class="relative flex h-0.5 items-center justify-center"
        >
          <div
            class="absolute bottom-0 left-0 top-0 rounded-lg bg-teal-400"
            style="width: -11%;"
          />
        </div>
      </div>
    </div>
    <div
      class="flex h-2/3 w-full flex-col items-center justify-between px-3 "
    >
      <div>
        <h1
          class="text-2xl font-extrabold text-gray-600"
        >
          Question 1?
        </h1>
        <span
          class="my-10 flex justify-center"
        >
          <button
            class="mx-3 inline-flex w-1/2 items-center justify-center rounded-full px-8 py-4 text-sm font-bold shadow-sm ring-1 ring-inset ring-gray-300
                        bg-white text-teal-400"
          >
            Yes
          </button>
          <button
            class="mx-3 inline-flex w-1/2 items-center justify-center rounded-full px-8 py-4 text-sm font-bold shadow-sm ring-1 ring-inset ring-gray-300
                        bg-white text-teal-400"
          >
            No
          </button>
        </span>
      </div>
      <button
        class="my-4 inline-flex w-full items-center justify-center  bg-gray-300 rounded px-4 py-2 font-bold text-white"
        disabled=""
      >
        Next
        <svg
          aria-hidden="true"
          class="h-5 w-5"
          data-slot="icon"
          fill="currentColor"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M6.22 4.22a.75.75 0 0 1 1.06 0l3.25 3.25a.75.75 0 0 1 0 1.06l-3.25 3.25a.75.75 0 0 1-1.06-1.06L8.94 8 6.22 5.28a.75.75 0 0 1 0-1.06Z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
